garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+        pass
+
+    @property
+    def name(self):
+        """Name for this Model."""
+        pass
+
+    @property
+    def input(self):
+        """Tensor input of the Model."""
+        pass
+
+    @property
+    def output(self):
+        """Tensor output of the Model."""
+        pass
+
+    @property
+    def inputs(self):
+        """Tensor inputs of the Model."""
+        pass
+
+    @property
+    def outputs(self):
+        """Tensor outputs of the Model."""
+        pass
+
+    @property
+    def parameters(self):
+        """Parameters of the Model."""
+        pass
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set parameters of the Model."""
+        pass
+
+
+class AutoPickableModel(BaseModel, metaclass=abc.ABCMeta):
+    """
+    Abstract class for models with automatic pickling.
+
+    This class follows the BaseModel API and handles pickling automatically.
+    It provides both self.init_spec and self.parameters for pickling, which
+    are the arguments and weights of the Model. When unpickled, a new model
+    will be reconstructed with self.init_spec as arguments and load the weights
+    from self.parameters.
+
+    The target weights should be assigned to self._default_parameters during
+    pickling, so that the newly created model can check if target weights exist
+    or not.
+    """
+
+    def __init__(self):
+        self._name = type(self).__name__  # name default to class
+        self._inputs = None
+        self._outputs = None
+        self._default_parameters = None
+
+    @property
+    def init_spec(self):
+        """
+        Model specification.
+
+        Positional and named arguments (args and kwargs) which, when passed
+        to `__init__`, will construct this model.
+        """
+        pass
+
+    @property
+    def name(self):
+        """Name of the model."""
+        return self._name
+
+    @property
+    def input(self):
+        """Tensor input of the Model."""
+        return self._inputs[0]
+
+    @property
+    def output(self):
+        """Tensor output of the Model."""
+        return self._outputs[0]
+
+    @property
+    def inputs(self):
+        """Tensor inputs of the Model."""
+        return self._inputs
+
+    @property
+    def outputs(self):
+        """Tensor outputs of the Model."""
+        return self._outputs
+
+    @property
+    def parameters(self):
+        """Parameters of the Model."""
+        raise NotImplementedError
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        raise NotImplementedError
+
+    def __getstate__(self):
+        """Object.__getstate__."""
+        args, kwargs = self.init_spec
+        return {'args': args, 'kwargs': kwargs, 'parameters': self.parameters}
+
+    def __setstate__(self, data):
+        """Object.__setstate__."""
+        # TODO(krzentner): There has got to be a better way.
+        out = type(self)(*data['args'], **data['kwargs'])
```

Comment:
@krzentner

Reply:
I don't really know of a better way of doing this. What really concerns me here isn't the use of `type()`, but that if the model implementer forgets a parameter in `init_spec()`, then we could silently lose data. Since this would be very hard to debug (and we promised not to lose data), this would be very bad.

Reply:
hmm very good point.

i still think quick_init() is too magical though. 

can we use something like your builder class idea instead?

Reply:
https://github.com/rlworkgroup/proposals/pull/3

Reply:
Yeah, I have also clearly articulated why I don't like `quick_init()`, but this is worse. This allows losing data for the same reason as `Serializable`, but *also* if a field is forgotten in `init_params`.

What stands in the way of doing this:
 - Make `build` a factory method which produces a subgraph.
 - Make Model naively serializable (since it would basically just contain the same things as `init_params()` would return).